runelite-client/src/main/java/net/runelite/client/plugins/microbot/tempoross/TemporossScript.java

```diff
@@ -0,0 +1,662 @@
+
+            if (action.equals("Tether")) {
+                if (Rs2GameObject.interact(tether, action)) {
+                    log(action + "ing");
+                    sleepUntil(() -> TemporossPlugin.isTethered == TemporossPlugin.incomingWave, 3500);
+                }
+            }
+            if (action.equals("Untether")) {
+                    log(action + "ing");
+                    sleepUntil(() -> TemporossPlugin.isTethered == TemporossPlugin.incomingWave, 3500);
+                    sleep(600,1200);
+
+            }
+        }
+    }
+
+    private void handleStateLoop() {
+        temporossPool = Rs2Npc.getNpc(NpcID.SPIRIT_POOL);
+        NPC inactivePool = Rs2Npc.getNpc(NpcID.INACTIVE_SPIRIT_POOL);
+        NPC doubleFishingSpot = Rs2Npc.getNpc(NpcID.FISHING_SPOT_10569);
+
+        if (TemporossScript.state == State.INITIAL_COOK && doubleFishingSpot != null) {
+            TemporossScript.state = TemporossScript.state.next;
+        }
+
+        if(TemporossScript.ENERGY < 30 && State.getAllFish() > 6 && !temporossConfig.solo() && TemporossScript.state != State.ATTACK_TEMPOROSS) {
+            TemporossScript.state = State.EMERGENCY_FILL;
+        }
+        if (TemporossScript.ENERGY == 0 && !temporossConfig.solo() && TemporossScript.state != State.ATTACK_TEMPOROSS && TemporossScript.state != State.INITIAL_CATCH) {
+            TemporossScript.state = State.ATTACK_TEMPOROSS;
+        }
+        if (temporossPool != null && TemporossScript.state != State.SECOND_FILL && TemporossScript.state != State.ATTACK_TEMPOROSS && TemporossScript.ENERGY < 94) {
+            TemporossScript.state = State.ATTACK_TEMPOROSS;
+        }
+        if (state.isComplete()) {
+            state = state.next;
+            if (state == null) {
+                log("Script looped, resetting to third catch");
+                state = State.THIRD_CATCH;
+            }
+        }
+    }
+
+    private void handleMainLoop() {
+        final Predicate<NPC> filterDangerousNPCs = npc -> !inCloud(npc.getWorldLocation(),1);
+        switch (state) {
+            case INITIAL_CATCH:
+            case SECOND_CATCH:
+            case THIRD_CATCH:
+                isFilling = false;
+                if (inCloud(Microbot.getClient().getLocalPlayer().getWorldLocation(),5)) {
+                    GameObject cloud = sortedClouds.stream()
+                            .findFirst()
+                            .orElse(null);
+                    Rs2Walker.walkNextToInstance(cloud);
+                    Rs2Player.waitForWalking();
+                    return;
+                }
+
+                NPC fishSpot = Rs2Npc.getNpcs(NpcID.FISHING_SPOT_10569)
+                        .filter(npc -> npc.getWorldLocation().distanceTo(workArea.rangePoint) <= 20)
+                        .filter(filterDangerousNPCs)
+                        .findFirst()
+                        .orElse(null);
+
+                if (fishSpot == null) {
+                    fishSpot = Rs2Npc.getNpcs()
```

Comment:
I think there is no need to query npcs twice if you already done it above? I think this block can be simplified.

below is some pseudo code to show you what i mean

```
fishSpot = Rs2Npc.getNpcs()
        .filter(npc -> npc.getId() == NpcID.FISHING_SPOT_10569 || npc.getId() == NpcID.FISHING_SPOT_10568 || npc.getId() == NpcID.FISHING_SPOT_10565)
        .filter(npc -> npc.getWorldLocation().distanceTo(workArea.rangePoint) <= 20)
        .filter(filterDangerousNPCs)
        .findFirst()
        .orElse(null);
```
        
```
        if (fishSpot != null) {
    if (!TemporossPlugin.isTethered && (!Rs2Player.isInteracting() || !Microbot.getClient().getLocalPlayer().getInteracting().equals(fishSpot))) {
        Rs2Camera.turnTo(fishSpot);
        Rs2Npc.interact(fishSpot, "Harpoon");
        log("Interacting with " + (fishSpot.getId() == NpcID.FISHING_SPOT_10569 ? "double" : "single") + " fish spot");
        Rs2Player.waitForWalking(2000);
    }
} else {
    if (!TemporossPlugin.isTethered && sortedFires.isEmpty()) {
        LocalPoint localPoint = LocalPoint.fromWorld(Microbot.getClient().getTopLevelWorldView(), workArea.totemPoint);
        Rs2Camera.turnTo(localPoint);
        Rs2Walker.walkFastLocal(localPoint);
        log("Can't find the fish spot, walking to the totem pole");
        Rs2Player.waitForWalking(2000);
    }
}
```

Reply:
Done